rabbit/rabbittransport.go

```diff
@@ -0,0 +1,426 @@
+	return serviceName
+}
+
+func (t *rabbitTransport) Tomb() *tomb.Tomb {
+	return t.tomb
+}
+
+func (t *rabbitTransport) connection() *RabbitConnection {
+	t.connM.RLock()
+	defer t.connM.RUnlock()
+	return t.conn
+}
+
+func (t *rabbitTransport) Ready() <-chan struct{} {
+	t.connM.RLock()
+	defer t.connM.RUnlock()
+	return t.connReady
+}
+
+func (t *rabbitTransport) killListeners() {
+	t.listenersM.RLock()
+	ts := make([]*tomb.Tomb, 0, len(t.listeners))
+	for _, t := range t.listeners {
+		t.Killf("Listeners killed")
+		ts = append(ts, t)
+	}
+	t.listenersM.RUnlock()
+	for _, t := range ts {
+		t.Wait()
+	}
+}
+
+func (t *rabbitTransport) StopListening(serviceName string) bool {
+	t.listenersM.RLock()
+	tm, ok := t.listeners[t.deliveryChan(serviceName)]
+	if ok {
+		tm.Killf("Stopped listening")
+	}
+	t.listenersM.RUnlock()
+	if ok {
+		tm.Wait()
+		return true
+	}
+	return false
+}
+
+func (t *rabbitTransport) Listen(serviceName string, rc chan<- message.Request) error {
+	tm := &tomb.Tomb{}
+	cn := t.deliveryChan(serviceName)
+	t.listenersM.Lock()
+	if _, ok := t.listeners[cn]; ok {
+		t.listenersM.Unlock()
+		return transport.ErrAlreadyListening
+	} else {
+		t.listeners[cn] = tm
+		t.listenersM.Unlock()
+	}
+
+	// Used to convey a connection error to the caller (we block until listening has begun)
+	errChan := make(chan error)
+
+	tm.Go(func() error {
+		timeout := time.NewTimer(connectTimeout)
+		defer func() {
+			timeout.Stop()
+			t.listenersM.Lock()
+			defer t.listenersM.Unlock()
+			delete(t.listeners, cn)
+			close(rc)
+			close(errChan)
+			log.Debugf("[Typhon:RabbitTransport] Listener %s stopped", cn)
+		}()
+
+		select {
+		case <-t.tomb.Dying():
+			return nil
+		case <-tm.Dying():
+			return nil
+		case <-timeout.C:
+			errChan <- transport.ErrTimeout
+			return nil
+		case <-t.Ready():
+		}
+
+		deliveryChan, rabbitChannel, err := t.connection().Consume(cn)
+		if err != nil {
+			log.Warnf("[Typhon:RabbitTransport] Failed to consume from %s: %v", cn, err)
+			errChan <- err
+			return nil
+		}
+		defer rabbitChannel.Close()
+		errChan <- nil
+		log.Infof("[Typhon:RabbitTransport] Listening on %s…", cn)
+
+		for {
+			select {
+			case <-t.tomb.Dying():
+				return nil
+
+			case <-tm.Dying():
+				return nil
+
+			case delivery, ok := <-deliveryChan:
+				if !ok {
+					log.Warnf("[Typhon:RabbitTransport] Delivery channel closed; stopping listener %s", cn)
+					return nil
+				}
+				go t.handleReqDelivery(delivery, rc)
+			}
+		}
+	})
+	return <-errChan
+}
+
+func (t *rabbitTransport) logId(delivery amqp.Delivery) string {
+	return fmt.Sprintf("%s[%s]", delivery.RoutingKey, delivery.CorrelationId)
+}
+
+func (t *rabbitTransport) Respond(req message.Request, rsp message.Response) error {
+	headers := rsp.Headers()
+	headers["Content-Encoding"] = "response"
+	headers["Service"] = rsp.Service()
+	headers["Endpoint"] = rsp.Endpoint()
+
+	timeout := time.NewTimer(respondTimeout)
```

Comment:
Ideally this would be configurable at some point, and potentially even be aware of the current request deadline if/when we put these in. We can live with 10s for this PR though!